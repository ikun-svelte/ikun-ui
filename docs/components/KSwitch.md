---
title: KSwitch
lang: en-US
---

# KSwitch

Use Switch to store contents.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/switch
```

```bash [yarn]
yarn add @ikun-ui/switch
```

```bash [npm]
npm install @ikun-ui/switch
```

:::

## Basic usage

<demo src="switch/basic.svelte" github="Switch"></demo>

## Disabled switch

Use `disabled` attribute to determine whether a switch is disabled.

<demo src="switch/disabled.svelte" github="Switch"></demo>

## Loading switch

Use `disabled` attribute to determine whether a switch is loading.

<demo src="switch/loading.svelte" github="Switch"></demo>

## Custom color switch

<demo src="switch/color.svelte" github="Switch"></demo>

## Custom switch value

<demo src="switch/value.svelte" github="Switch"></demo>

## Custom switch render

<demo src="switch/render.svelte" github="Switch"></demo>

## Switch Props

| Name           | Type                        | Default | Description                  |
| -------------- | --------------------------- | ------- | ---------------------------- |
| value          | `boolean`                   | `false` | Binding value                |
| unCheckedValue | `string / number / boolean` | `false` | UnChecked state switch value |
| checkedValue   | `string / number / boolean` | `true`  | Checked state switch value   |
| disabled       | `boolean`                   | `false` | Disabled the switch          |
| loading        | `boolean`                   | `-`     | Loading state switch         |
| unCheckedColor | `string`                    | `-`     | unChecked state switch color |
| checkedColor   | `string`                    | `-`     | Checked state switch color   |
| cls            | `string`                    | `-`     | Additional class             |
| attrs          | `Record<string, string>`    | `{}`    | Additional attributes        |

## Switch Events

| Name        | Description                                              | Type                    |
| ----------- | -------------------------------------------------------- | ----------------------- |
| click       | Triggered when the `switch` is clicked                   | `(value: Event)=> void` |
| updateValue | Triggered when the value in the `switch` binding changes | `(value: Event)=> void` |
| change      | Triggered when the value in the `switch` binding changes | `(value: Event)=> void` |

## Switch Slots

| Name            | Description                                    |
| --------------- | ---------------------------------------------- |
| unCheckedRender | Customize content of the uncheked state switch |
| checkedRender   | Customize content of the cheked state switch   |
