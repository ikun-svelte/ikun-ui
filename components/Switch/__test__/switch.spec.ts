import { tick } from 'svelte';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import KSwitch from '../src';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.restoreAllMocks();
});

describe('Test: KSwitch', () => {
	test('props: value', async () => {
		const value = true;
		const instance = new KSwitch({
			target: host,
			props: {
				value
			}
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-switch__checked')).toBe(true);
		instance.$set({
			value: false
		});
		await tick();
		expect(host.innerHTML.includes('k-switch__unchecked')).toBe(true);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: unCheckedValue and checkedValue', async () => {
		let value = '微笑';
		const instance = new KSwitch({
			target: host,
			props: {
				value,
				unCheckedValue: '哭泣',
				checkedValue: '微笑'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		instance.$on('updateValue', (v) => {
			value = v.detail;
			instance.$set({
				value,
				unCheckedValue: '哭泣',
				checkedValue: '微笑'
			});
		});
		const switchElm = host.getElementsByTagName('span')[0];
		switchElm.click();
		await tick();
		expect(value).toBe('哭泣');
		switchElm.click();
		await tick();
		expect(value).toBe('微笑');
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: disabled', async () => {
		const instance = new KSwitch({
			target: host,
			props: {
				disabled: true
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('k-switch__disabled')).toBe(true);
		instance.$set({
			disabled: false
		});
		await tick();
		expect(host.innerHTML.includes('k-switch__disabled')).toBe(false);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: loading', async () => {
		const instance = new KSwitch({
			target: host,
			props: {
				loading: true
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('k-switch-loading')).toBe(true);
		instance.$set({
			loading: false
		});
		await tick();
		expect(host.innerHTML.includes('k-switch-loading')).toBe(false);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: unCheckedColor and checkedColor', async () => {
		const instance = new KSwitch({
			target: host,
			props: {
				value: false,
				unCheckedColor: '#520',
				checkedColor: '#1314'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('#520')).toBe(true);
		instance.$set({
			value: true
		});
		await tick();
		expect(host.innerHTML.includes('#1314')).toBe(true);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: size', async () => {
		const instance = new KSwitch({
			target: host,
			props: {
				size: 'sm'
			}
		});
		await tick();
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-switch--sm')).toBeTruthy();
		expect(host.innerHTML.includes('k-switch-circle--sm')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();

		instance.$set({ size: 'md' });
		await tick();
		expect(host.innerHTML.includes('k-switch--md')).toBeTruthy();
		expect(host.innerHTML.includes('k-switch-circle--md')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();

		instance.$set({ size: 'lg' });
		await tick();
		expect(host.innerHTML.includes('k-switch--lg')).toBeTruthy();
		expect(host.innerHTML.includes('k-switch-circle--lg')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: cls', async () => {
		const instance = new KSwitch({
			target: host,
			props: {
				cls: '关关雎鸠,在河之洲'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('关关雎鸠,在河之洲')).toBe(true);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: attrs', async () => {
		const instance = new KSwitch({
			target: host,
			props: {
				attrs: {
					you: 'world'
				}
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.children[0].getAttribute('you')).toBe('world');
		expect(host.innerHTML).matchSnapshot();
	});

	test('event: should trigger click event', async () => {
		const mockFn = vi.fn();
		const instance = new KSwitch({
			target: host,
			props: {
				value: true
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		instance.$on('click', () => {
			mockFn();
		});
		const switchElm = host.getElementsByClassName('k-switch--base')[0] as HTMLDivElement;
		switchElm.click();
		await tick();
		await vi.advanceTimersByTimeAsync(350);
		expect(mockFn).toBeCalled();
	});

	test('event: should trigger updateValue event', async () => {
		const updateEvent = vi.fn();
		const instance = new KSwitch({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		instance.$on('updateValue', () => {
			updateEvent();
		});
		const switchElm = host.getElementsByClassName('k-switch--base')[0] as HTMLDivElement;
		switchElm.click();
		await tick();
		expect(updateEvent).toBeCalled();
	});

	test('event: should trigger change event', async () => {
		const changeEvent = vi.fn();
		const instance = new KSwitch({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		instance.$on('change', () => {
			changeEvent();
		});
		const switchElm = host.getElementsByClassName('k-switch--base')[0] as HTMLDivElement;
		switchElm.click();
		await tick();
		expect(changeEvent).toBeCalled();
	});

	test('event: should not trigger click & updateValue & change event when disabled', async () => {
		const eventFn = vi.fn();
		const instance = new KSwitch({
			target: host,
			props: {
				disabled: true
			}
		});
		expect(instance).toBeTruthy();
		instance.$on('updateValue', eventFn);
		instance.$on('change', eventFn);
		instance.$on('click', eventFn);
		await tick();
		const switchElm = host.getElementsByClassName('k-switch--base')[0] as HTMLDivElement;
		switchElm.click();
		expect(eventFn).not.toBeCalled();
	});
});
