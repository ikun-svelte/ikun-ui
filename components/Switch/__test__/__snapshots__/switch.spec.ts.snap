// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KSwitch > props: attrs 1`] = `"<div class=\\"k-switch k-switch--base k-switch--md k-switch__unchecked\\" aria-hidden=\\"true\\" you=\\"world\\"><span class=\\"k-switch-tx__unchecked k-switch-tx__unchecked--md\\"></span> <span class=\\"k-switch-circle k-switch-circle--md\\"></span> </div>"`;

exports[`Test: KSwitch > props: cls 1`] = `"<div class=\\"k-switch k-switch--base k-switch--md k-switch__unchecked 关关雎鸠,在河之洲\\" aria-hidden=\\"true\\"><span class=\\"k-switch-tx__unchecked k-switch-tx__unchecked--md\\"></span> <span class=\\"k-switch-circle k-switch-circle--md\\"></span> </div>"`;

exports[`Test: KSwitch > props: disabled 1`] = `"<div class=\\"k-switch k-switch--base k-switch--md k-switch__unchecked\\" aria-hidden=\\"true\\"><span class=\\"k-switch-tx__unchecked k-switch-tx__unchecked--md\\"></span> <span class=\\"k-switch-circle k-switch-circle--md\\"></span> </div>"`;

exports[`Test: KSwitch > props: loading 1`] = `"<div class=\\"k-switch k-switch--base k-switch--md k-switch__unchecked\\" aria-hidden=\\"true\\"><span class=\\"k-switch-tx__unchecked k-switch-tx__unchecked--md\\"></span> <span class=\\"k-switch-circle k-switch-circle--md\\"></span> </div>"`;

exports[`Test: KSwitch > props: size 1`] = `"<div class=\\"k-switch k-switch--base k-switch--sm k-switch__unchecked\\" aria-hidden=\\"true\\"><span class=\\"k-switch-tx__unchecked k-switch-tx__unchecked--sm\\"></span> <span class=\\"k-switch-circle k-switch-circle--sm\\"></span> </div>"`;

exports[`Test: KSwitch > props: size 2`] = `"<div class=\\"k-switch k-switch--base k-switch--md k-switch__unchecked\\" aria-hidden=\\"true\\"><span class=\\"k-switch-tx__unchecked k-switch-tx__unchecked--md\\"></span> <span class=\\"k-switch-circle k-switch-circle--md\\"></span> </div>"`;

exports[`Test: KSwitch > props: size 3`] = `"<div class=\\"k-switch k-switch--base k-switch--lg k-switch__unchecked\\" aria-hidden=\\"true\\"><span class=\\"k-switch-tx__unchecked k-switch-tx__unchecked--lg\\"></span> <span class=\\"k-switch-circle k-switch-circle--lg\\"></span> </div>"`;

exports[`Test: KSwitch > props: unCheckedColor and checkedColor 1`] = `"<div class=\\"k-switch k-switch--base k-switch--md k-switch__checked #1314\\" aria-hidden=\\"true\\"> <span class=\\"k-switch-circle k-switch-circle--md\\"></span> <span class=\\"k-switch-tx__checked k-switch-tx__checked--md\\"></span></div>"`;

exports[`Test: KSwitch > props: unCheckedValue and checkedValue 1`] = `"<div class=\\"k-switch k-switch--base k-switch--md k-switch__checked k-switch-tra\\" aria-hidden=\\"true\\"> <span class=\\"k-switch-circle k-switch-circle--md\\" style=\\"right: 1px;\\"></span> <span class=\\"k-switch-tx__checked k-switch-tx__checked--md\\"></span></div>"`;

exports[`Test: KSwitch > props: value 1`] = `"<div class=\\"k-switch k-switch--base k-switch--md k-switch__unchecked\\" aria-hidden=\\"true\\"><span class=\\"k-switch-tx__unchecked k-switch-tx__unchecked--md\\"></span> <span class=\\"k-switch-circle k-switch-circle--md\\"></span> </div>"`;
